src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs

```diff
@@ -391,14 +398,29 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
             // If none of those are the case then we can just take an early out.
 
             ArrayBuilder<LocalSymbol> temporariesBuilder = ArrayBuilder<LocalSymbol>.GetInstance();
-            rewrittenArguments = _factory.MakeTempsForDiscardArguments(rewrittenArguments, temporariesBuilder);
             ImmutableArray<ParameterSymbol> parameters = methodOrIndexer.GetParameters();
 
-            if (CanSkipRewriting(rewrittenArguments, methodOrIndexer, expanded, argsToParamsOpt, invokedAsExtensionMethod, false, out var isComReceiver))
+            if (CanSkipRewriting(
+                    arguments,
+                    methodOrIndexer,
+                    expanded,
+                    argsToParamsOpt,
+                    invokedAsExtensionMethod,
+                    ignoreComReceiver: false,
+                    argumentsWillBeReused,
+                    out var isComReceiver,
+                    out var interpolatedStringHandlerArgumentIndex))
             {
-                temps = temporariesBuilder.ToImmutableAndFree();
                 argumentRefKindsOpt = GetEffectiveArgumentRefKinds(argumentRefKindsOpt, parameters);
 
+                ImmutableArray<BoundExpression> rewrittenArguments = arguments.SelectAsArray(
+                    static (argument, context) => argument is BoundDiscardExpression discard
```

Comment:
Consider expanding this into a loop, to improve readability.

Reply:
I actually don't think the expanded version is more readable. It's more than a foreach loop: I will also need to check the arguments length for the 0 case, allocate a builder, and manage it. This is much more concise.